Convert robot_ws package to ROS2

timrobotson opened the pull request:
Convert all the launch files, package.xml, and nodes to be ROS2 dashing compatible. 

The launch files are now equivalent to the ROS1 versions, so you can specify LAUNCH_ID or ROS_AWS_REGION environment variables which will do equivilant overrides and you can specify the same `aws_metrics_namespace` or `log_group_name` parameters, and they fall back to the config file if unset (using sample app config file if it's set, or CE config file otherwise).

Requires these PR's to be merged to work correctly:
https://github.com/aws-robotics/cloudwatchlogs-ros2/pull/6
https://github.com/aws-robotics/cloudwatchmetrics-ros2/pull/8
https://github.com/aws-robotics/health-metrics-collector-ros2/pull/3

I'll publish simulation_ws changes in a separate PR to make them not so overwhelming. simulation_ws changes are still WIP. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

timrobotson committed:
Initial convert of files using ros_upgrader.py

timrobotson committed:
Convert nodes and launch files

- Convert all nodes to ROS2
- Fix launch files to include other launch files

timrobotson committed:
Make nodes compile and run correctly

- Add new required dependencies for turtlebot to .rosinstall
- Remove CMakeLists as it's no longer needed
- Make launch files include other launch filres correctly
- Move nodes to the new src/cloudwatch_robot folder.

timrobotson committed:
Remove unneccessary dependencies from .rosinstall

timrobotson committed:
Fix suscribers/publishers to be correct order of params

timrobotson committed:
Fix launch files including other launch files

timrobotson committed:
Include health/metrics/logs collectors in launch files

- In the monitoring launch file include the launch descriptions for
health metrics collector, cloudwatch metrics collector and cloudwatch
logger.
- Pass the same arguments to these included files as previously

timrobotson committed:
Revert simulation_ws changes for now

timrobotson committed:
Pull in region and launch_id from environment variables and pass to nodes

AAlon commented on robot_ws/.rosinstall:
```diff
@@ -1,8 +1,13 @@
+- git: {local-name: src/deps/health-metrics-collector-ros2, uri: 'https://github.com/aws-robotics/health-metrics-collector-ros2', version: 'master'}
+- git: {local-name: src/deps/monitoringmessages-ros2, uri: 'https://github.com/aws-robotics/monitoringmessages-ros2', version: 'master'}
+- git: {local-name: src/deps/turtlebot3, uri: 'https://github.com/ROBOTIS-GIT/turtlebot3.git', version: 'ros2'}
+- git: {local-name: src/deps/turtlebot3_msgs, uri: 'https://github.com/ROBOTIS-GIT/turtlebot3_msgs.git', version: 'ros2'}
```
Can remove turtlebot3_msgs, DynamixelSDK, hls_lfcd_lds_driver (released to apt)

AAlon commented on robot_ws/src/cloudwatch_robot/launch/monitoring.launch.py:
```diff
@@ -0,0 +1,87 @@
+        ),
+        launch.actions.DeclareLaunchArgument(
+            name='metrics_node_name',
+            default_value="metrics_ros2"
```
Not a fan of metrics_ros2, logger_ros2 - let's be consistent with the ROS1 implementation as much as possible (except when there's a good reason not to). I believe we use the names `cloudwatch_metrics_collector` and `cloudwatch_logger` there.

timrobotson committed:
Make launch files equivilant to ROS1 launch files

- With monitoring launch file you can now specify LAUNCH_ID
environment variable and it will correctly modify the metrics
namespace or log group name as it did in ROS1
- With monitoring launch file you can now specify AWS_REGION environment
variable  which will overwrite the config file.
- log_group_name and metrics_namespace can now be overwritten and
otherwise default to the value in the configuration file.
- Update sample configuration files to be ROS2 compatible and have
required arguments.

timrobotson committed:
Use updated health_metric_collector and cloudwatch_metrics_collector packages

AAlon commented:
Nice! Have we verified that this works as expected in regards to parameter passing?
A good way to test it might be to launch the app with aws_region parameter set to eu-west-1 (anything other than us-west-2) and verify the nodes' parameters, following up with verification in CloudWatch console (that the metrics/logs were indeed emitted to eu-west-1).

AAlon commented:
To get the CI build passing could you please adjust the travis.yml like we did in https://github.com/aws-robotics/aws-robomaker-sample-application-helloworld/blob/ros2/.travis.yml ? You can put a COLCON_IGNORE in simulation_ws if that isn't ready yet.

ryanewel commented on robot_ws/src/cloudwatch_robot/package.xml:
```diff
@@ -1,39 +1,33 @@
-
+  <license>MIT</license>
+  <buildtool_depend>ament_python</buildtool_depend>
+  <buildtool_depend>rosidl_default_generators</buildtool_depend>
```
Is this necessary? This package isn't building any message types. (Not 100% sure on this)

timrobotson replied:
Fixed

timrobotson committed:
Update travis, delete old launch files, fix bugs

- Update travis.yml file for ROS2 and disable pushing to pipelines for
now
- Get turtlebot dependencies from apt instead of git cloning
- Delete all old launch files

timrobotson commented:
> Nice! Have we verified that this works as expected in regards to parameter passing?

Yep tested with `ros2 launch cloudwatch_robot monitoring.launch.py aws_region:=ap-northeast-1` and it correct sent all logs and metrics to that region. 

> To get the CI build passing could you please adjust the travis.yml 

done

AAlon commented on robot_ws/src/cloudwatch_robot/launch/await_commands.launch.py:
```diff
@@ -0,0 +1,24 @@
+        ),
+        launch.actions.IncludeLaunchDescription(
+            launch.launch_description_sources.PythonLaunchDescriptionSource(
+                [get_package_share_directory('cloudwatch_robot'), '/launch/monitoring.launch.py']
```
os.path.join instead of `/`
(occurs in multiple places so ditto for all other as well)

timrobotson replied:
I tried doing this and it doesn't work, I think because `get_package_share_directory` doesn't resolve correctly until later, so when doing the join in this spot it resolves to an empty string. I tried both with and without surrounding braces (so array or string it doesn't work)

timrobotson replied:
Upon further investigation the issue is that when using `os.path.join` if the second argument starts with / it treats it as an absolute path and ignores the first argument. I'd rather keep it with a / in the front to be consistent with how ROS1 functions (with a leading slash before launch file).

AAlon replied:
I'm not sure I understand the reasoning. The slash signifies a path and the platform-independent, Pythonic way to build paths is using os.path.join.

AAlon commented on robot_ws/src/cloudwatch_robot/launch/deploy_rotate.launch.py:
```diff
@@ -0,0 +1,30 @@
+import sys
+
+import launch
+import launch_rpos.actions
```
meant `launch_ros`?

AAlon commented on robot_ws/src/cloudwatch_robot/package.xml:
```diff
@@ -1,39 +1,32 @@
-  <depend>roscpp</depend>
-
+  <license>MIT</license>
+  <buildtool_depend>ament_python</buildtool_depend>
```
I don't think ament_python is a buildtool option. we probably don't need a buildtool here? (I've seen some python packages use `ament_package`)

AAlon commented on robot_ws/src/cloudwatch_robot/package.xml:
```diff
@@ -1,39 +1,32 @@
+<?xml version='1.0' encoding='utf-8'?>
+<package format="3">
   <name>cloudwatch_robot</name>
   <version>1.1.0</version>
```
bump version to 2.0.0

AAlon commented on robot_ws/src/cloudwatch_robot/package.xml:
```diff
@@ -1,39 +1,32 @@
   <depend>turtlebot3</depend>
   <depend>turtlebot3_msgs</depend>
+  <depend>builtin_interfaces</depend>
+  <exec_depend>rosidl_default_runtime</exec_depend>
```
do we need this exec depend?
And add exec depend on `launch` and `launch_ros`.

AAlon commented on robot_ws/src/cloudwatch_robot/setup.py:
```diff
@@ -1,13 +1,47 @@
-    package_dir={'': 'src'}
+setup(
+    name=package_name,
+    version='1.0.0',
```
Is this supposed to match the version in package.xml?

AAlon reviewed (commented):
Great seeing this progress, mostly minor comments

AAlon committed:
Use apt testing repo

AAlon committed:
Use turtlebot3 and cartographer_ros from apt

timrobotson committed:
Fix version, typos, package requirements

- Major version bump to 2.0
- Fix small typos in python scripts
- Fix package.xml requirements
- Fix health_metrics_config to have a float instead of int

timrobotson committed:
Use os.path.join for launch file paths

- Use os.path.join for safety and to be consistent with other ROS2
launch files.

AAlon commented on robot_ws/src/cloudwatch_robot/launch/await_commands.launch.py:
```diff
@@ -0,0 +1,24 @@
+        ),
+        launch.actions.IncludeLaunchDescription(
+            launch.launch_description_sources.PythonLaunchDescriptionSource(
+                os.path.join(get_package_share_directory('cloudwatch_robot'), 'launch/monitoring.launch.py')
```
Use os.path.join to build the entire path:
`os.path.join(get_package_share_directory('cloudwatch_robot'), 'launch', 'monitoring.launch.py')`

timrobotson committed:
Split launch paths into individual files/folders

AAlon commented on robot_ws/src/cloudwatch_robot/launch/monitoring.launch.py:
```diff
@@ -0,0 +1,121 @@
+
+
+def generate_launch_description():
+    default_metrics_config = f"{get_package_share_directory('cloudwatch_robot')}/config/cloudwatch_metrics_config.yaml"
```
Doesn't need to be an f-string + use os.path.join :-)

timrobotson committed:
Change fstrings to os.path.join
